### ISSUE TITLE
Autotype error reports error "Could not open X display for autotyping"

### ISSUE THREAD

**Issue by igor2x** [flathub/org.pwsafe.pwsafe#7]
Installed Password Safe v1.18.2 as flatpak from Flathub on Ubuntu 23.10.

Right click on entry and select Perform Auto Type. It reports error: "Autotype error. There was an error autotyping. Could not open X display for autotyping."

Probably some security restriction by flatpak. But I am having similar problem using deb package, but in this case no action is performed and no error message. Because not the same behavior deb vs. flatpak, I am reporting issue on flatpak bug tracker.

**Comment by igor2x** [comment]
Just a quick update. On my Ubuntu I use Wayland as display server, because switching to X-server on Ubuntu logon and Ubuntu freezes. In my case only Wayland works and X-server does not. I can't confirm or dismiss if this "Autotype" is purely Wayland issue.

**Comment by igor2x** [comment]
@jjberk, can you please copy above message to original bug report at https://github.com/pwsafe/pwsafe/issues/1105
because this is different issue. And then delete your post here, to reduce the clutter. I will respond on original issue after you copy/paste.

**Comment by wohler** [comment]
After getting flatpak Password Safe 1.19.1 running on my Debian GNU/Linux 12 (bookworm) system, I got this error when I tried autotype for the first time as well. I'm running Xwayland at home. Have you found any workarounds in the meantime?

I'm not seeing this error on my Red Hat 8 system at work. The X server is X/Xwayland (TBD when I get into work).

**Comment by igor2x** [comment]
No, I haven't found a work around for this issue. I just double click on entry in Password Safe and then move to the window I want to have password and paste the entry.

---

If in _Flatseal_ for Passwrod Safe I disable "Wayland" and "Fallback to X11" like this:
![image](https://github.com/user-attachments/assets/ac96b021-86d9-4a44-b190-4cb362c6f3cc)

this should force flatpak to run using Xwayland.

If I start up Password Safe, autotype still does not work. Interesting is, it switches to the last viewed window (some kind of a progress), but password is not copied.

**Comment by wohler** [comment]
In case it is helpful, the Debian package for Password Safe was able to do autotype. It's version 1.17.

**Comment by wohler** [comment]
Thanks for the Flatseal tip. I found that X connectivity issues with Flatpak is common. One post suggested turning off socket=wayland and leaving on socket=x11 and socket=fallback-x11. I did this, restarted Password Safe, and autotype worked!

I reset the settings in Flatseal and added --nosocket=wayland to the Exec entry in my ~/.local/share/applications/org.pwsafe.pwsafe.desktop file. That works too.

Is this something can be added to the flatpak manifest?

**Comment by igor2x** [comment]
My system: Ubuntu 24.04 using Password Safe v1.19.1/flatpak on Wayland

@wohler, I don' know how you have done this, but I have tried:
1. _Reset_ button in _Flatseal_ for _Password Safe_ application.
2. From terminal: `flatpak run --nosocket=wayland org.pwsafe.pwsafe`
3. Typed in master password.
4. Switch to _Text Editor_ and switch back to _Password Safe_.
5. Right mouse click on entry and _Perform Auto Type_.
6. No password is copied to _Text Editor_ and _Password Safe_ is minimized.

Just to make extra sure, can you execute command: `echo $XDG_SESSION_TYPE` to see if you are really using _wayland_.

---

> Is this something can be added to the flatpak manifest?

Yes. If we look at the [manifest file](https://github.com/flathub/org.pwsafe.pwsafe/blob/master/org.pwsafe.pwsafe.yml)  there is section:
```
finish-args:
  - --socket=fallback-x11
  - --socket=wayland
```
and changing last line to `--nosocket=wayland` should do the trick.

**Comment by wohler** [comment]
```
$ echo $XDG_SESSION_TYPE
wayland
```
I noticed that you had the socket=x11 option turned off in your examples above. For completeness, my socket configuration is
`--socket=x11 --nosocket=wayland --socket=fallback-x11`. Does that help?

**Comment by igor2x** [comment]
Unfortunately it does now work for me.

I tried:
1. I have opened _Flatseal_ to make sure all settings are set to default.
2. Then I have run:
`flatpak run --socket=x11 --nosocket=wayland --socket=fallback-x11 org.pwsafe.pwsafe`
3. Clicked on _Text Editor_ and Alt+Tab to Password Safe and right click on entry and selected _Perform Auto Type_ focus is switches to _Text Editor_ but unfortunately password is not copied. I don't know why it does not work for me.

I have double checked:
- Double click on password and paste in _Text Editor_ and password is correctly pasted.
- Right click and _Edit Entry_ in _Additional_ tab _Autotype_ field is empty, so it should honor _Manage | Options | Miscellaneuos | Autotype section | Default Autotype string_ is set to `\p`
- Right click and _Edit Entry_ in _Additional_ tab _Autotype_ field I have entered `\p`

---

> It seems ironic that I have to turn off wayland to work with wayland...>

A lot of GUI applications on Linux were written for X-window and do not work well on Wayland. At Wayland project they have done double implementation:
- GUI application can call native Wayland calls.
- GUI application can call X-window calls using so called XWayland system and those calls are then translated to Wayland calls.

I think XWayland is good idea, so applications can migrate step by step to Wayland. If on the other side XWayland was not implemented, I suppose Wayland would never get live in Linux distributions, because there is just too much legacy applications.

**Comment by wohler** [comment]
I tried autotype with my editor (Emacs), and text was not entered there either. Try it in your browser.

**Comment by igor2x** [comment]
I tried with Firefox browser for hotmail.com and gmail.com and in both cases password is not copied after _Perform Auto Type_. If I wait after few seconds Password Safe reappears with notification:
![image](https://github.com/user-attachments/assets/8d4f6d78-0b45-497a-9f45-e4e55c1b8f38)
This notification is just like a notification that is displayed when Password Safe is started.

**Comment by wohler** [comment]
Bummer. This isn't as general a workaround as I (or Rony I suppose) had hoped.

If you click on that notification, I suspect you'll see the Can't connect to X display error.

**Comment by froggie2523** [comment]
The app has been updated to provide users with information on how to manually enable "full X11 access" to make the AutoType feature work on Wayland systems. See https://github.com/pwsafe/pwsafe/pull/1943 for more details. 
The next Flatpak release will include these changes. I think it's OK to close this one.
